### ISSUE TITLE
.Net: Stepwise Planner: Planner fails to mark final answer as [Final Answer]

### ISSUE THREAD

**Issue by mtechvance** [microsoft/semantic-kernel#2466]
**Describe the bug**
With some goals, the stepwise planner will eventually have a thought that should be the final answer but fails to mark it as such. It continues to repeat a thought that should be the final answer until it runs out of iterations. My specific scenario is analyzing code from our repository based on a user chat. I have noticed that this only occurs for larger code files. Smaller code files usually end up having a final answer.

Example scratch pad from the planner showing the repetition

> This was my previous work (but they haven't seen any of it! They only see what I return as final answer):
> [THOUGHT] To determine if AminoStartup configures Swagger, I need to search for relevant code references in the MTech Systems' agricultural software. I will use the CodeSearchSkill.SearchCodeByReference function to search for "AminoStartup" and "Swagger" in the codebase.
> [ACTION] {"action": "CodeSearchSkill.SearchCodeByReference","action_variables": {"input":"","searchText":"AminoStartup Swagger"}}
> [OBSERVATION] [
> 	"**file path removed**"
> ]
> [THOUGHT] I found a file named "AminoStartup.cs" that might contain the configuration for Swagger. I will now use the CodeSearchSkill.GetFileContent function to get the content of this file and check if it configures Swagger.
> [ACTION] {"action": "CodeSearchSkill.GetFileContent","action_variables": {"input":"","filePath":"**file path removed**"}}
> [OBSERVATION] 
> [FILE CONTENT]
> **FileContent would be here**
> [END FILE CONTENT]
> [THOUGHT] Yes, AminoStartup configures Swagger in the MTech Systems' agricultural software. I found a file named "AminoStartup.cs" in the codebase, and upon examining its content, I found that it includes the necessary configuration for Swagger.
> [THOUGHT] Yes, AminoStartup configures Swagger in the MTech Systems' agricultural software. I found a file named "AminoStartup.cs" in the codebase, and upon examining its content, I found that it includes the necessary configuration for Swagger.

**To Reproduce**
1. Download repro project.
2. Enter your endpoint and key in the Program.cs file.
3. Open CodeSearchSkill and paste the contents of a large code file in the variable FileContentToReturn
4. Run application and observe the stepwise planner scratch pad

**Expected behavior**
The first complete answer returned about the file should be marked as final answer

**Platform**
 - OS: Windows
 - IDE: Visual Studio
 - Language: C#
 - Source: NuGet 0.19.230804.2-preview

[ConsoleApp2.zip](https://github.com/microsoft/semantic-kernel/files/12346056/ConsoleApp2.zip)

**Comment by lemillermicrosoft** [comment]
I have tried to resolve this in #2504 -- if you were able to supply an example code file for repro I can do a better job confirming.

**Comment by mtechvance** [comment]
I tested this morning by pasting the code from stepwiseplanner.cs into the CodeSearchSkill of my console project. Screen shots below of the code pasted in and the planner spinning on the same thought. Ignore the 'AminoStartup' text.

![image](https://github.com/microsoft/semantic-kernel/assets/95925478/8bb59003-b622-441b-ab62-b1664a4b544f)

![image](https://github.com/microsoft/semantic-kernel/assets/95925478/0ff804ba-9089-4a53-8e28-7931cb2a7ce8)

**Comment by mtechvance** [comment]
@lemillermicrosoft,
I was doing some testing with the new PR you linked and it does look promising for this issue's use case. One other issue I found is that the action RegEx does not match if there are curly braces '{}' anywhere within the action variables. See snippet below. I have a curly brace in the 'comment' property.

```
void Main()
{
	Regex s_actionRegex = new(@"\[ACTION\][^{}]*({(?:[^{}]*{[^{}]*})*[^{}]*})", RegexOptions.Singleline | RegexOptions.IgnoreCase);

	var text = @"[ACTION]{
  ""action"": ""RepositorySkill.PushChangesToBranch"",
  ""action_variables"": {
    ""branchName"": ""myBranchName"",
    ""comment"": ""{MyComment""
  }
}
";

	s_actionRegex.Match(text).Dump();
}


```

**Comment by lemillermicrosoft** [comment]
Yeah another known issue #2553

**Comment by lemillermicrosoft** [comment]
Added a fix to #2504 for the Action parsing issues.

**Comment by nacharya1** [comment]
@lemillermicrosoft : can you update target dates for this issue, dates are in the past.
